rust/src/mysql/mysql.rs

```diff
@@ -0,0 +1,1131 @@
+                Ok((i, MysqlBEMessage::Response(resp)))
+            }
+
+            MysqlStateProgress::StmtFetchReceived
+            | MysqlStateProgress::StmtFetchResponseContinue => {
+                let (i, resp) = parse_stmt_fetch_response(i)?;
+                Ok((i, MysqlBEMessage::Response(resp)))
+            }
+
+            MysqlStateProgress::FieldListReceived => {
+                let (i, resp) = parse_field_list_response(i)?;
+                Ok((i, MysqlBEMessage::Response(resp)))
+            }
+            MysqlStateProgress::StatisticsReceived => {
+                let (i, resp) = parse_statistics_response(i)?;
+                Ok((i, MysqlBEMessage::Response(resp)))
+            }
+            MysqlStateProgress::ChangeUserReceived => {
+                let (i, resp) = parse_change_user_response(i)?;
+                Ok((i, MysqlBEMessage::Response(resp)))
+            }
+
+            _ => {
+                let (i, resp) = parse_response(i)?;
+                Ok((i, MysqlBEMessage::Response(resp)))
+            }
+        }
+    }
+
+    fn invalid_state_resp(&self) -> bool {
+        use MysqlStateProgress::*;
+        self.state_progress == CommandResponseReceived
+            || self.state_progress == StmtCloseReceived
+            || self.state_progress == StmtPrepareResponseReceived
+            || self.state_progress == StmtExecResponseReceived
+    }
+
+    pub fn parse_response(&mut self, flow: *const Flow, i: &[u8]) -> AppLayerResult {
+        // We're not interested in empty responses.
+        if i.is_empty() {
+            return AppLayerResult::ok();
+        }
+
+        if self.response_gap {
+            if !probe(i).is_ok() {
+                SCLogDebug!("Suricata interprets there's a gap in the response");
+                return AppLayerResult::ok();
+            }
+
+            // It seems we're in sync with a message header, clear gap state and keep parsing.
+            self.response_gap = false;
+        }
+
+        let mut start = i;
+
+        while !start.is_empty() {
+            if self.state_progress == MysqlStateProgress::Finished || self.invalid_state_resp() {
+                return AppLayerResult::ok();
+            }
+            match MysqlState::state_based_resp_parsing(
+                self.state_progress,
+                start,
+                self.client_flags,
+            ) {
+                Ok((rem, response)) => {
+                    start = rem;
+
+                    SCLogDebug!("Response is {:?}", &response);
+                    if let Some(state) = self.response_next_state(response, flow) {
+                        self.state_progress = state;
+                    }
+                    let tx_completed = self.is_tx_completed();
+                    if let Some(tx) = self.find_or_create_tx() {
+                        if tx_completed {
+                            tx.complete = true;
+                        }
+                    }
```

Comment:
Should we not create a new transaction with the response if ` self.find_or_create_tx() ` returns None ?

Reply:
Sorry, I don't understand the question.